File: November/23/hindex2.py
# LeetCode 275: H-Index II
# Explanation:
# 1. We need the largest h such that citations[n - h] >= h.
# 2. Use binary search on h from 0..n.
# 
# Time Complexity: O(log n)
# Space Complexity: O(1)

class Solution:
    def hIndex(self, citations: list[int]) -> int:
        n = len(citations)
        left, right = 0, n

        while left < right:
            mid = (left + right + 1) // 2  # proposed h
            if citations[n - mid] >= mid:
                left = mid
            else:
                right = mid - 1

        return left

File: November/23/hindex2.cpp
// Explanation:
// 1. citations is sorted ascending.
// 2. We binary search for minimum index i where citations[i] >= n - i.
// 3. The h-index is n - i.
// Time Complexity: O(log n)
// Space Complexity: O(1)

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        int l = 0, r = n - 1;

        while (l <= r) {
            int m = (l + r) / 2;
            if (citations[m] >= n - m)
                r = m - 1;
            else
                l = m + 1;
        }
        return n - l;
    }
};

// Example usage:
// Solution sol;
// cout << sol.hIndex(vector<int>{0,1,3,5,6}); // 3
// cout << sol.hIndex(vector<int>{1,2,100});   // 2
// cout << sol.hIndex(vector<int>{0});         // 0
